src/lualib/Promise.ts

```diff
@@ -0,0 +1,335 @@
+declare function error( value:string, level?:number ): never;
+const logError: any = print;
+
+let runAtNextTick: ( func: () => void ) => void = func => func();
+function setRunAtNextTickFunction(  func: (f: () => void ) => void ): void {
+  runAtNextTick = func;
+}
+
+enum PromiseState {
+  PENDING = 0,
+  FULFILLED = 1,
+  REJECTED = 2,
+}
+
+interface Deferred< T, TResult > {
+  promise: PromiseImpl<TResult>;
+  onFulfilled?: (( this: void, value: T) => TResult) | undefined;
+  onRejected?: (( this: void, reason: any) => TResult) | undefined;
+}
+
+class PromiseImpl<T = never> {
+
+  public static resolve<T = never>(): PromiseImpl<T>;
+  public static resolve<T = never>( value: T | PromiseImpl<T> ): PromiseImpl<T>;
+  public static resolve<T = never>( value?: T | PromiseImpl<T> ): PromiseImpl<T> {
+    const p = new PromiseImpl<T>();
+    PromiseImpl.PromiseResolutionProcedure( p, value as T );
+    return p;
+  }
+
+  public static reject( reason: any ): PromiseImpl<never> {
+    const p = new PromiseImpl();
+    p.reject( reason );
+    return p;
+  }
+
+  protected static create<T = undefined>(): PromiseImpl<T> {
```

Comment:
It's not a part of promise standard and doesn't look like it's used by something internally. Should it be removed?